### PR TITLE
Custom Data Attributes in the toolbar

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

- PRs over issues, so this is a bit raw and needs feedback and or patches
- Adds the following area to the project setup screen: 
image
- Then if you actually need to create a selector on an element with one of those IDs, this whitelisted `data` attribute will take precedence. For example: 

The idea came from talking to and observing users. Some of them do instrument their app with `data-attr` attributes, but many have some existing system in place, where the attribute has a slightly different name, like `data-posthog-id`. This PR adds support for these.

Notes and next steps:
- We can't blatantly use all data attributes, as many of them actually store, uhm, data. Like input field values, whatever angular-specific crap, etc.
- The UX of the setup page could absolutely be improved. Saving a CSV list like that is hacky
- Documentation needs to be updated.
- We could autodetect and suggest some custom attributes like we do for `app_urls` now.
- We could also let people add custom `data-attributes` to the list right in the toolbar, e.g. when clicking on button that has `data-custom-blabla` that's not already in the list, we suggest "Is `data-custom-blabla` a useful thing for your site?" or so...
- If a custom attribute is in the list of matched elements, we could show it on the events table. "Clicked button" is less interesting than "clicked [data-attr='homepage-cta-button']". This would help at least one EE customer.

Overall, all feedback is greatly welcome! Especially regarding naming things and database columns :)

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

**Comment by macobo** [comment]
:+1: for tacking this. Related issues: https://github.com/PostHog/posthog/issues/2140, https://github.com/PostHog/posthog/issues/2362

The solution is still incomplete though and won't work for a large set of usecases where the data attribute is arbitrary - e.g. `data-test-click-button`, etc. This is the more common usecase - the way we use `data-attr` is completely specific to posthog and while this is the step in the right direction it doesn't really solve it.

Could we use a regex here instead?

**Comment by mariusandra** [comment]
Sure! Perhaps just a simpler `*` for catch all (and `?` for catch one, maybe?) will give most benefits in an easier way?

**Comment by macobo** [comment]
Sounds perfect. The case here is indeed to basically whitelist some prefix. :)

File: posthog/models/team.py [hunk 1]
```diff
@@ -99,6 +99,9 @@ class Team(models.Model):
     is_demo: models.BooleanField = models.BooleanField(default=False)
     test_account_filters: JSONField = JSONField(default=list)
     timezone: models.CharField = models.CharField(max_length=240, choices=TIMEZONES, default="UTC")
+    important_data_attributes: ArrayField = ArrayField(
```
Comment (macobo) [inline posthog/models/team.py:102]: Note: This same model has used `JSONField` before for this w/ the event_names etc columns. I think it's worth being consistent.

**Comment by mariusandra** [comment]
Wildcard support added! Just "*" to keep things simple.

image

File: posthog/models/team.py (hunk 1 above)
Comment (mariusandra) [reply posthog/models/team.py]: Makes sense. I also renamed it from `important_data_attributes` to just `data_attributes`. `important` is not a very informative prefix.

**Comment by mariusandra** [comment]
Now brave explorers can also write `data-*` and have any data attribute take strong precedence. I guess some teams might even find that useful.

What should be the default value for this? Currently I have it set to `data-attr`, but does that make sense? Alternatives are `` (nothing) and `data-*` (all data attributes, filter down yourself if needed)

**Comment by mariusandra** [comment]
Finally, one problem I'll leave untackled in this PR is that if you update your project settings, the toolbar won't see the changes. You'll have to re-launch or re-authorize the toolbar to get the new attributes. The data is only sent with the authorisation. This is slightly hostile/unexpected UX. Right now with this (semi?) secret "temporaryToken", we can't do any API access from within the toolbar, so a broader fix is needed.

**Comment by yakkomajuri** [comment]
Woop, thanks for using the tag @macobo !

File: frontend/src/scenes/project/Settings/DataAttributes.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                be used when creating actions to match unique elements on your pages.
+            </p>
+            <p>
+                For example, when creating an action on your CTA button, the best selector could be something like:{' '}
+                <code>div &gt; form &gt; button:nth-child(2)</code>. However all buttons in your app have a{' '}
+                <code>data-custom-id</code> attribute. If you whitelist it here, the selector for your button will
+                instead be <code>button[data-custom-id='cta-button']</code>.
+            </p>
+            <div>
+                <Input
```
Comment (macobo) [inline frontend/src/scenes/project/Settings/DataAttributes.tsx:35]: WDYT about using SelectGradientOverflow here to do the multi-input rather than doing this comma-based? See https://github.com/PostHog/posthog/pull/3756/files#diff-f180e14d42a7636d030be25ac92466e7a890170edfe246d5b9f0f63c46121634R80 for an example.

File: frontend/src/toolbar/toolbarLogic.ts [hunk 3]
```diff
@@ -26,6 +26,14 @@ export const toolbarLogic = kea<toolbarLogicType>({
         actionId: [props.actionId || null, { logout: () => null, clearUserIntent: () => null }],
         userIntent: [props.userIntent || null, { logout: () => null, clearUserIntent: () => null }],
         buttonVisible: [true, { showButton: () => true, hideButton: () => false, logout: () => false }],
+        dataAttributes: [
+            ((typeof props.dataAttributes === 'string'
```
Comment (macobo) [inline frontend/src/toolbar/toolbarLogic.ts:30]: Q: Why is this needed? (I see a similar joining in the backend, but this is weird)

Also, Perhaps a cleaner way to do this is with a selecror rather than reducer.

File: frontend/src/scenes/project/Settings/DataAttributes.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            <p>
+                Specify a comma-separated list of{' '}
+                <a
+                    href="https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes"
+                    rel="noreferrer noopener"
+                >
+                    data attributes
+                </a>{' '}
+                used in your app. For example: <code>data-attr, data-custom-id, data-myref</code>. These attributes will
+                be used when creating actions to match unique elements on your pages.
```
Comment (macobo) [inline frontend/src/scenes/project/Settings/DataAttributes.tsx:26]: Suggestion: `be used when using the toolbar and defining actions to match unique elements on your pages`

File: frontend/src/scenes/project/Settings/DataAttributes.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            </p>
+            <p>
+                For example, when creating an action on your CTA button, the best selector could be something like:{' '}
+                <code>div &gt; form &gt; button:nth-child(2)</code>. However all buttons in your app have a{' '}
+                <code>data-custom-id</code> attribute. If you whitelist it here, the selector for your button will
+                instead be <code>button[data-custom-id='cta-button']</code>.
```
Comment (macobo) [inline frontend/src/scenes/project/Settings/DataAttributes.tsx:32]: Suggestion. `You can use * as a wildcard`.

File: frontend/src/scenes/project/Settings/DataAttributes.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            </p>
+            <div>
+                <Input
+                    placeholder="data-attr, ..."
```
Comment (macobo) [inline frontend/src/scenes/project/Settings/DataAttributes.tsx:36]: Suggestion: `data-analytics*` for the default + placeholder

File: frontend/src/toolbar/utils.ts [hunk 7]
```diff
@@ -30,17 +30,20 @@ export function getSafeText(el: HTMLElement): string {
     return elText
 }
 
-export function elementToQuery(element: HTMLElement): string | undefined {
+export function elementToQuery(element: HTMLElement, dataAttributes: string[]): string | undefined {
     if (!element) {
         return
     }
+    if (!simmer) {
```
Comment (macobo) [inline frontend/src/toolbar/utils.ts:37]: Q: Why the lazy init?

File: posthog/migrations/0139_team_data_attributes.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 3.0.11 on 2021-03-26 05:34
+
+import django.contrib.postgres.fields
+from django.db import migrations, models
+
+
+def set_default_data_attributes(apps, schema_editor):
+    Team = apps.get_model("posthog", "Team")
+    Team.objects.update(data_attributes=["data-attr"])
```
Comment (macobo) [inline posthog/migrations/0139_team_data_attributes.py:9]: Suggestion: `data-analytics*` as well.

**Review by macobo — COMMENTED** [review]
Added some thoughts.

File: frontend/src/toolbar/utils.ts (hunk 7 above)
Comment (mariusandra) [reply frontend/src/toolbar/utils.ts:37]: Cause we don't have the dataAttributes when the JS is evaluated on the root level.

File: frontend/src/scenes/project/Settings/DataAttributes.tsx (hunk 2 above)
Comment (mariusandra) [reply frontend/src/scenes/project/Settings/DataAttributes.tsx]: Good idea! Though since we're not autocompleting anything here, I just used `Select` instead of `SelectGradientOverflow`.

File: frontend/src/toolbar/toolbarLogic.ts (hunk 3 above)
Comment (mariusandra) [reply frontend/src/toolbar/toolbarLogic.ts:30]: Indeed, was not needed.

**Comment by mariusandra** [comment]
Bunch of feedback addressed. I'm not 100% convinced of `data-analytics*` though. The places where we've reported it, we've used `data-attr` and changing this without warning will break someone's setup. Ours as well. I've been in concat with some people who are using it actively for example. I could add it *in addition* to `data-attr`... 🤔

**Comment by mariusandra** [comment]
One last fix to the migration (defaults added also to all new users, not just while migration), ready for another look.

**Comment by macobo** [comment]
> I could add it in addition to data-attr... thinking

Yep, that's my suggestion because of that reasoning. It demonstrates the wildcard functionality in a tangible way as well.

**Comment by mariusandra** [comment]
Hm.. Thinking of it, I'd actually just leave it at `data-attr` for now. Defaults in something like this are notoriously hard to change after, e.g. after many users start using `data-analytics*`. What is `analytics` :)? Perhaps instead of that after a small team chat we'd actually decide to use `data-ph-*` or something else? So I'd just defer this decision.

**Review by macobo — APPROVED** [review]
Sorry for the delay, let's ship
